src/main/java/com/github/manolo8/darkbot/gui/zones/ZoneEditor.java

```diff
@@ -85,88 +60,145 @@ public void mouseMoved(MouseEvent e) {
 
     void setup(Main main, java.util.Map<Integer, ZoneInfo> zonesByMap) {
         super.setup(main);
-        if (mapChange != null) main.mapManager.mapChange.remove(mapChange);
-        main.mapManager.mapChange.add(mapChange = map -> {
-            zoneInfo = zonesByMap.computeIfAbsent(map.id, id -> new ZoneInfo(config.BOT_SETTINGS.OTHER.ZONE_RESOLUTION));
-            ZoneEditor.this.repaint();
-        });
-        zoneInfo = zonesByMap.get(main.hero.map.id);
+        this.zonesDrawer = new ZoneEditorDrawer(main.pluginAPI, this, zonesByMap);
     }
 
     @Override
-    protected void paintComponent(Graphics g) {
-        Graphics2D g2 = setupDraw(g);
+    protected void onPaint() {
+        zonesDrawer.onDraw(mapGraphics);
+    }
+
+    public static class ZoneEditorDrawer extends SafetiesDisplay.SafetiesDisplayDrawer<ZoneEditor> implements Listener {
```

Comment:
I've explained it more in safeties display, but i don't understand the need to create an inner-class for these editor drawers, they simply act as a way to display the map on their own, they don't need to create a drawer because the order is hard-coded and we don't plan on supporting anything rendering in between, these editors stay as-is.

Also, it seems like a very weird way to do this for ZoneEditor to extend SafetiesDisplay, what does one have to do with the other semantically? they shouldn't inherit eachother